src/mbgl/gl/render_target_gl.cpp

```diff
@@ -0,0 +1,85 @@
+#include <mbgl/gl/render_target_gl.hpp>
+#include <mbgl/gl/context.hpp>
+#include <mbgl/gl/defines.hpp>
+#include <mbgl/gl/framebuffer.hpp>
+#include <mbgl/platform/gl_functions.hpp>
+#include <mbgl/renderer/layer_tweaker.hpp>
+#include <mbgl/renderer/paint_parameters.hpp>
+#include <mbgl/renderer/render_orchestrator.hpp>
+#include <mbgl/renderer/render_static_data.hpp>
+#include <mbgl/renderer/render_tree.hpp>
+#include <mbgl/util/logging.hpp>
+
+#include <cassert>
+
+namespace mbgl {
+namespace gl {
+
+using namespace platform;
+
+RenderTargetGL::RenderTargetGL(Context& context, const Size size, const gfx::TextureChannelDataType type)
+    : glContext(context) {
+    texture = glContext.createTexture2D();
+    texture->setSize(size);
+    texture->setFormat(gfx::TexturePixelType::RGBA, type);
+    texture->setSamplerConfiguration(
+        {gfx::TextureFilterType::Linear, gfx::TextureWrapType::Clamp, gfx::TextureWrapType::Clamp});
+}
+
+RenderTargetGL::~RenderTargetGL() {
+    if (framebuffer) {
+        MBGL_CHECK_ERROR(glDeleteFramebuffers(1, &framebuffer->get()));
+        framebuffer.reset();
+    }
+}
+
+void RenderTargetGL::upload(gfx::UploadPass& uploadPass) {
+    observeLayerGroups(([&](LayerGroupBase& layerGroup) { layerGroup.upload(uploadPass); }));
+}
+
+void RenderTargetGL::render(RenderOrchestrator& orchestrator,
+                            const RenderTree& renderTree,
+                            PaintParameters& parameters) {
+    if (!framebuffer) {
+        texture->create();
+        framebuffer = std::make_shared<UniqueFramebuffer>(glContext.createFramebuffer(*texture));
+    }
+
+    glContext.bindFramebuffer = *framebuffer;
+    glContext.activeTextureUnit = 0;
+    glContext.scissorTest = false;
+    glContext.viewport = {0, 0, texture->getSize()};
+    glContext.clear(Color{0.0f, 0.0f, 0.0f, 1.0f}, {}, {});
+
+    // Run layer tweakers to update any dynamic elements
+    observeLayerGroups([&](LayerGroupBase& layerGroup) {
+        if (layerGroup.getLayerTweaker()) {
+            layerGroup.getLayerTweaker()->execute(layerGroup, renderTree, parameters);
+        }
+    });
+
+    // draw layer groups, opaque pass
```

Comment:
This does a lot of what `Renderer::Impl::render()` does, we should consider modularization of that logic and break it out of `Impl::render()` so we can also call it here.

Reply:
Yeah, it reminded me a bit of the way we do render targets in WhirlyGlobe.  We eventually mostly consolidated that for the default render target (the screen) and other render targets (other textures we draw to).  But it wasn't trivial.

I think it's worth opening an issue and having someone spend a bit of time on it.  It's something I'd like to see fleshed out in MapLibre.

Reply:
Yes, we will need to optimize this.